### PR TITLE
Switch Stackdriver sink to grpc-based Stackdriver API

### PR DESCRIPTION
**Pull request by kawych**
This PR together with #1907 lowers Heapster maximum resource usage by 60% CPU and 25% memory (tested over 2 hours on GCE, 100-node cluster with 3k running user pods).

### CODE REVIEW

**Comment by kawych** [comment]
@pongad
Can you take a look?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 1]
```diff
@@ -751,48 +751,51 @@ func (sink *StackdriverSink) getNodeResourceLabels(labels map[string]string) map
...
-func legacyCreateTimeSeries(resourceLabels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
+func legacyCreateTimeSeries(resourceLabels map[string]string, metadata *metricMetadata, point *monitoringpb.Point) *monitoringpb.TimeSeries {
 	return createTimeSeries("gke_container", resourceLabels, metadata, point)
 }
 
-func createTimeSeries(resource string, resourceLabels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
-	return &sd_api.TimeSeries{
-		Metric: &sd_api.Metric{
+func createTimeSeries(resource string, resourceLabels map[string]string, metadata *metricMetadata, point *monitoringpb.Point) *monitoringpb.TimeSeries {
+	//monitoringpb.
```
Comment (igorpeshansky) [inline metrics/sinks/stackdriver/stackdriver.go]: ?

Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 2]
```diff
@@ -15,39 +15,41 @@
...
 	"github.com/golang/glog"
+	google_protobuf2 "github.com/golang/protobuf/ptypes/timestamp"
+	"github.com/googleapis/gax-go"
 	"github.com/prometheus/client_golang/prometheus"
-	"golang.org/x/oauth2"
-	"golang.org/x/oauth2/google"
-	"google.golang.org/api/googleapi"
-	sd_api "google.golang.org/api/monitoring/v3"
+	google_api5 "google.golang.org/genproto/googleapis/api/metric"
+	google_api4 "google.golang.org/genproto/googleapis/api/monitoredres"
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: Maybe rename these imports?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 3]
```diff
@@ -102,8 +104,8 @@ func (sink *StackdriverSink) Stop() {
 }
 
 func (sink *StackdriverSink) processMetrics(metricValues map[string]core.MetricValue,
-	timestamp time.Time, labels map[string]string, collectionStartTime time.Time, entityCreateTime time.Time) []*sd_api.TimeSeries {
-	timeseries := make([]*sd_api.TimeSeries, 0)
+	timestamp time.Time, labels map[string]string, collectionStartTime time.Time, entityCreateTime time.Time) []*monitoringpb.TimeSeries {
+	timeseries := make([]*monitoringpb.TimeSeries, 0)
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: `var timeseries []*monitoringpb.TimeSeries`? That way empty arrays don't allocate.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 4]
```diff
@@ -268,7 +270,7 @@ func (sink *StackdriverSink) requestSender(reqQueue chan *sd_api.CreateTimeSerie
 	}
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: Isn't this loop just
```go
for req := range reqQueue {sink.sendOneRequest(req)}
```
?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 5]
```diff
@@ -277,11 +279,11 @@ func marshalRequestAndLog(printer func([]byte), req *sd_api.CreateTimeSeriesRequ
...
 }
 
-func (sink *StackdriverSink) sendOneRequest(req *sd_api.CreateTimeSeriesRequest) {
+func (sink *StackdriverSink) sendOneRequest(req *monitoringpb.CreateTimeSeriesRequest) {
 	startTime := time.Now()
-	empty, err := sink.stackdriverClient.Projects.TimeSeries.Create(fullProjectName(sink.project), req).Do()
+	err := sink.stackdriverClient.CreateTimeSeries(context.Background(), req, gax.WithGRPCOptions())
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: No need to write `gax.WithGRPCOptions()` if you don't pass options.
This function is used for passing more options to gRPC layer.

File: metrics/sinks/stackdriver/stackdriver.go [hunk 6]
```diff
@@ -801,6 +803,9 @@ func fullProjectName(name string) string {
 	return fmt.Sprintf("projects/%s", name)
 }
 
-func getReq() *sd_api.CreateTimeSeriesRequest {
-	return &sd_api.CreateTimeSeriesRequest{TimeSeries: make([]*sd_api.TimeSeries, 0)}
+func getReq(project string) *monitoringpb.CreateTimeSeriesRequest {
+	return &monitoringpb.CreateTimeSeriesRequest{
+		TimeSeries: make([]*monitoringpb.TimeSeries, 0),
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: `nil`?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 7]
```diff
@@ -751,48 +751,50 @@ func (sink *StackdriverSink) getNodeResourceLabels(labels map[string]string) map
...
-func (sink *StackdriverSink) intPoint(endTime time.Time, startTime time.Time, value int64) *sd_api.Point {
-	return &sd_api.Point{
-		Interval: &sd_api.TimeInterval{
-			EndTime:   endTime.Format(time.RFC3339),
-			StartTime: startTime.Format(time.RFC3339),
+func (sink *StackdriverSink) intPoint(endTime time.Time, startTime time.Time, value int64) *monitoringpb.Point {
+	return &monitoringpb.Point{
+		Interval: &monitoringpb.TimeInterval{
+			EndTime:   &google_protobuf2.Timestamp{Seconds: endTime.Unix(), Nanos: int32(endTime.Nanosecond())},
+			StartTime: &google_protobuf2.Timestamp{Seconds: startTime.Unix(), Nanos: int32(startTime.Nanosecond())},
```
Comment (pongad) [inline metrics/sinks/stackdriver/stackdriver.go]: Consider using [this](https://godoc.org/github.com/golang/protobuf/ptypes#TimestampProto)

**Review by pongad — COMMENTED** [review]
A couple of nits, but I think it generally looks good.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 2 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 3 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 4 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 5 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

File: metrics/sinks/stackdriver/stackdriver.go (hunk 7 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: I think it actually returns different type.

File: metrics/sinks/stackdriver/stackdriver.go (hunk 6 above)
Comment (kawych) [reply metrics/sinks/stackdriver/stackdriver.go]: Done

**Review by piosz — APPROVED** [review]
/lgtm

**Pull request merged**